images/virtualization-artifact/pkg/controller/service/snapshot_service.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2024 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	vsv1 "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	virtv1 "kubevirt.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/deckhouse/virtualization-controller/pkg/sdk/framework/helper"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vmcondition"
+)
+
+const virtualizationSubresourcePath = "/apis/subresources.virtualization.deckhouse.io/v1alpha2/namespaces/%s/virtualmachines/%s/%s"
+
+type SnapshotService struct {
+	restClient *rest.RESTClient
+	client     Client
+	protection *ProtectionService
+}
+
+func NewSnapshotService(restClient *rest.RESTClient, client Client, protection *ProtectionService) *SnapshotService {
+	return &SnapshotService{
+		restClient: restClient,
+		client:     client,
+		protection: protection,
+	}
+}
+
+func (s *SnapshotService) CanFreeze(vm *virtv2.VirtualMachine) bool {
+	if vm == nil || vm.Status.Phase != virtv2.MachineRunning {
+		return false
+	}
+
+	agentReady, _ := GetCondition(vmcondition.TypeAgentReady.String(), vm.Status.Conditions)
+	if agentReady.Status != metav1.ConditionTrue {
+		return false
+	}
+
+	filesystemReady, _ := GetCondition(vmcondition.TypeFilesystemReady.String(), vm.Status.Conditions)
+
+	return filesystemReady.Status == metav1.ConditionTrue
+}
+
+func (s *SnapshotService) Freeze(ctx context.Context, name, namespace string) error {
+	path := fmt.Sprintf(virtualizationSubresourcePath, namespace, name, "freeze")
+
+	body, err := json.Marshal(&virtv1.FreezeUnfreezeTimeout{
+		UnfreezeTimeout: &metav1.Duration{},
+	})
+	if err != nil {
+		return err
+	}
+
+	err = s.restClient.Put().AbsPath(path).Body(body).Do(ctx).Error()
```

Comment:
I think, you should use our kubeclient.  
Add new method Freeze for VirtualMachineInterface. For example: https://github.com/deckhouse/virtualization/blob/1f2f3807ae8ec49e0133da064b1e17aa7536da71/api/client/kubeclient/vm.go#L99